Tools/preflight.py

```diff
@@ -0,0 +1,282 @@
+# Enthought product code
+#
+# (C) Copyright 2010-2022 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This file and its contents are confidential information and NOT open source.
+# Distribution is prohibited.
+
+"""
+    Script for checking an app's Docker image for Edge compatibility.
+"""
+
+import logging
+import os
+import os.path as op
+import subprocess
+import sys
+import tempfile
+import time
+from contextlib import contextmanager
+from unittest import TestCase
+from urllib.parse import urlparse
+import shutil
+
+import click
+import requests
+from click import secho
+
+RETRIES = 5
+BACKOFF = 5
+
+
+@click.command()
+@click.argument("image")
+@click.option(
+    "--app-log", default="edge-app-log.txt", help="Write app log here on failure"
+)
+@click.option(
+    "--hub-log", default="edge-hub-log.txt", help="Write JupyterHub log here on failure"
+)
+def preflight(image, app_log, hub_log):
+    """Test an application's Docker image for compatibility with Edge."""
+
+    if "/" not in image or ":" not in image:
+        secho(f"Invalid image name '{image}'.", fg="red")
+        secho("Image name should be of the form 'example.com/org/repo:tag'")
+        sys.exit(1)
+
+    app_log = op.abspath(app_log)
+    hub_log = op.abspath(hub_log)
+
+    with run_jupyterhub(image, app_log, hub_log), requests.Session() as s:
+        with check("Get the login page"):
+            r = s.get("http://localhost:8000/hub/login")
+            r.raise_for_status()
+
+        with check("Perform login"):
+            response = s.post(
+                "http://localhost:8000/hub/login",
+                data={"username": "edgeuser", "password": "password"},
+                allow_redirects=True,
+            )
+            response.raise_for_status()
+
+        with check("Tell the spawner to start"):
+            response = s.get("http://localhost:8000/hub/spawn/edgeuser")
+            response.raise_for_status()
+
+        with check("Wait for the application to start"):
+            # Collect all redirects
+            redirects = []
+            for _ in range(RETRIES):
+                response = s.get("http://localhost:8000/user/edgeuser")
+                response.raise_for_status()
+                redirects = redirects + [urlparse(r.url).path for r in response.history]
+                if "spawn-pending" not in response.url:
+                    break
+                time.sleep(BACKOFF)
+            else:
+                raise AssertionError("Application failed to start")
+
+        with check("Fetch application home page"):
+            response = s.get("http://localhost:8000/user/edgeuser")
+            response.raise_for_status()
+            redirects = redirects + [urlparse(r.url).path for r in response.history]
+            if response.url != "http://localhost:8000/user/edgeuser/":
+                raise AssertionError("Application home page did not load")
```

Comment:
This fails for Panel, likely because Panel attempts to redirect `/user/edgeuser/` to `/user/edgeuser/app`. It may be possible to validate this by seeing if `/user/edgeuser/` is somewhere in the redirect history after the first redirect. This will also impact the `check("Check unknown users cannot access app")` test in the reverse fashion. (`/user/edgeuser/` will not be the response.url, therefore the check will pass when it shouldn't.)

Reply:
Good catch, I have updated both checks to use `url.startswith` instead.  Presumably the final URL must be located under the /user/edgeuser prefix.

Reply:
All of the URLs (oauth_start, error pages from nginx for 500, etc) will be under the prefix, so it should also check that the status code is within the 200 range